DIPs/DIP4242.md

```diff
@@ -0,0 +1,260 @@
+attributes of the function that receives it.
+However, this does not address the problem presented here: instead, it forces
+the user to require from the caller the attributes it supports.
+
+Doing so usually leads to attributes being faked, or the routine being unusable:
+for example, iterating over a collection might be a simple operation which is
+`pure` and `@nogc`, but requiring the delegate to also be is too limitating.
+
+## Description
+
+### Grammar
+
+The following changes to the grammar are proposed:
+```diff
+ StorageClass:
+     LinkageAttribute
+     AlignAttribute
+     deprecated
+     enum
+     static
+     extern
+     abstract
+     final
+     override
+     synchronized
+     auto
+     scope
+     const
+     immutable
+     inout
+     shared
+     __gshared
+     Property
+     nothrow
++    nothrow AttributeDeclDef
+     pure
++    pure AttributeDeclDef
+     ref
+ 
+ AtAttribute:
+     @ disable
+     @ nogc
++    @ nogc AttributeDeclDef
+     @ live
+     Property
+     @ safe
++    @ safe AttributeDeclDef
+     @ system
+     @ trusted
+     UserDefinedAttribute
+ 
++AttributeDeclDef:
++    ( * )
++    ( AttributeDeclDefs $(OPT) )
+
++AttributeDeclDefs:
++    AssignExpression
++    AssignExpression, AttributeDeclDefs ...
+```
+
+This syntax was choosen as it should feel familiar to the user,
+who can already encounter it when using UDAs.
+
+### Basic semantic rules
+
+If the argument list form is used, the argument must be either identifiers
+or integers. If identifiers, they must match the identifiers of one of
+the function's arguments, and this argument must be a delegate or function pointer.
+If integers, the value must be positive and at most one less than the function's
+arity (number of parameter), included. The value must be the 0-based index of the
+argument the attribute depends on. Likewise, this argumment needs to be a
+delegate or function pointer.
+To avoid special cases in meta-programming code, we follow the widespread practice
+of allowing empty lists and trailing commas.
+
+Hence, the following are valid:
+```D
+// Basic usage, with nothrow and @safe
+void func0(void delegate(int) sink) @safe(sink) nothrow(sink);
+// Empty argument list, equivalent to @safe
+void func1() @safe();
+// Equivalent to func0
+void func2(void delegate(int)) @safe(*) nothrow(*);
+// Equivalent to func0
+void func3(void delegate(int) arg) @safe(arg,) nothrow(*);
+// Equivalent to func1
+void func4(int) @safe(*);
+// Equivalent to func0
+void func3(void delegate(int) arg) @safe(0) nothrow(0,);
+```
+
+However, the following are not valid:
+```D
+// Argument name does not exists
+void err0(void delegate() sink) @safe(snk);
+// Integer out of bound
+void err1(void delegate() sink) @safe(1);
+// Argument is not a delegate or function pointer
+void err2(int arg) @safe(arg);
+// Argument is not a delegate or function pointer
+void err3(int arg) @safe(0);
```

Comment:
The index version is quite confusing because of implicit boolean conversion (this looks like `@safe(false)` to me)

Reply:
Didn't think of that. The reason it's in is to support meta programming, where you really don't want to have to mess with identifiers. Sometimes you declare a function with a type tuple and in this case having ADAs support indexes is quite handy.

Reply:
Couödn't the same be achieved by allowing `tuple[index]` to refer to the actual parameter inside of a tuple?

Reply:
I don't see how ? Or is `tuple` a keyword here ?

Reply:
Given the following function
```d
void foo(T...)(T args) {}
```

Your current proposal suggests a plain index to specifiy a single parameter:

```d
void foo(T...)(T args) @safe(0) {}
```

But this would be more expressive:

```d
void foo(T...)(T args) @safe(args[0]) {}
```

---

Also a plain index doesn't work well with multiple variadic template parameters:

```d
template bar(A...)
{
	void bar(B...)(A a, B b) @safe(1) {} // Might refer to an element of a or b
}
```

Reply:
Hum, those are good points. Index based ADAs are a special case for when an argument has no identifier. Since the grammar allows it, I figured there should be a way to handle it.

But even if that's the intended usage, the semantics should be made clearer. And the tuple example is quite neat and more expressive, I agree. Let me think this over for a bit.